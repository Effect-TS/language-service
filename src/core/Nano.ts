import * as Either from "effect/Either"
import { dual } from "effect/Function"
import type { TypeLambda } from "effect/HKT"
import * as Option from "effect/Option"
import * as Gen from "effect/Utils"

const NanoInternalSuccessProto = {
  _tag: "Right"
}

interface NanoInternalSuccess<A> {
  _tag: "Right"
  value: A
}

function makeInternalSuccess<A>(value: A): NanoExit<A, never> {
  const result = Object.create(NanoInternalSuccessProto)
  result.value = value
  return result
}

const NanoInternalFailureProto = {
  _tag: "Left"
}

interface NanoInternalFailure<E> {
  _tag: "Left"
  value: E
}

function makeInternalFailure<E>(value: E): NanoExit<never, E> {
  const result = Object.create(NanoInternalFailureProto)
  result.value = value
  return result
}

const NanoInternalDefectProto = {
  _tag: "Defect"
}

interface NanoInternalDefect {
  _tag: "Defect"
  value: unknown
}

function makeInternalDefect(value: unknown): NanoExit<never, never> {
  const result = Object.create(NanoInternalDefectProto)
  result.value = value
  return result
}

type NanoExit<A, E> =
  | NanoInternalSuccess<A>
  | NanoInternalFailure<E>
  | NanoInternalDefect

export class NanoDefectException {
  readonly _tag = "@effect/language-service/NanoDefectException"
  constructor(
    readonly message: unknown
  ) {}
}

export class NanoTag<R> {
  declare "~nano.requirements": R
  constructor(
    readonly key: string
  ) {}
}

export const Tag = <I = never>(identifier: string) => new NanoTag<I>(identifier)

type NanoContext<R = never> = {
  _R: R
  value: Record<string, unknown>
}

export const contextEmpty: NanoContext<never> = { value: {} } as any

export interface NanoIterator<T extends Nano<any, any, any>> {
  next(...args: ReadonlyArray<any>): IteratorResult<Gen.YieldWrap<T>, T["~nano.success"]>
}

export interface NanoTypeLambda extends TypeLambda {
  readonly type: Nano<this["Target"], this["Out1"], this["Out2"]>
}

/**
 * Nano is a Effect-like interface to run things.
 * It is not intended to be used by users in production.
 * It's only a mere tool to be used in the Effect dev-tools
 * to provide a familiar effect-like experience in envs
 * where using full blown Effect will cause an Effect-in-Effect issue.
 * It is supposed to be sync only and not stack-safe.
 * Thrown exceptions are catched and converted into defects,
 * so worst case scenario, you will get only standard typescript lsp.
 */
export interface Nano<out A = never, out E = never, out R = never> {
  readonly "~nano.success": A
  readonly "~nano.error": E
  readonly "~nano.requirements": R
  [Symbol.iterator](): NanoIterator<Nano<A, E, R>>
  run: (
    ctx: NanoContext<unknown>
  ) => NanoExit<A, E>
}

const Proto = {
  run: () => {},

  [Symbol.iterator]() {
    return new Gen.SingleShotGen(new Gen.YieldWrap(this))
  }
}

function make<A, E, R>(
  run: (
    ctx: NanoContext<unknown>
  ) => NanoExit<A, E>
): Nano<A, E, R> {
  const result = Object.create(Proto)
  result.run = run
  return result
}

export const unsafeRun = <A, E>(
  fa: Nano<A, E, never>
): Either.Either<A, E | NanoDefectException> => {
  const program = provideService(internalNanoCache, {})(fa)
  const result = program.run(contextEmpty)
  switch (result._tag) {
    case "Left":
      return Either.left(result.value)
    case "Defect":
      return Either.left(new NanoDefectException(result.value))
    case "Right":
      return Either.right(result.value)
  }
}

export const run = <A, E>(fa: Nano<A, E, never>): Either.Either<A, E | NanoDefectException> => {
  try {
    return unsafeRun(fa)
  } catch (e) {
    return Either.left(new NanoDefectException(e))
  }
}

export const succeed = <A>(value: A) => make<A, never, never>(() => makeInternalSuccess(value))
export const fail = <E>(value: E) => make<never, E, never>(() => makeInternalFailure(value))
export const sync = <A>(value: () => A) => make<A, never, never>(() => makeInternalSuccess(value()))
export const flatMap: {
  <A, B, E2, R2>(f: (a: A) => Nano<B, E2, R2>): <E, R>(fa: Nano<A, E, R>) => Nano<B, E | E2, R | R2>
  <A, E, R, B, E2, R2>(fa: Nano<A, E, R>, f: (a: A) => Nano<B, E2, R2>): Nano<B, E | E2, R | R2>
} = dual(2, <A, E, R, B, E2, R2>(
  fa: Nano<A, E, R>,
  f: (a: A) => Nano<B, E2, R2>
) =>
  make<B, E | E2, R | R2>((ctx) => {
    const result = fa.run(ctx)
    if (result._tag !== "Right") return result
    return f(result.value).run(ctx)
  }))

export const map: {
  <A, B>(f: (a: A) => B): <E, R>(fa: Nano<A, E, R>) => Nano<B, E, R>
  <A, E, R, B>(fa: Nano<A, E, R>, f: (a: A) => B): Nano<B, E, R>
} = dual(2, <A, E, R, B>(
  fa: Nano<A, E, R>,
  f: (a: A) => B
) =>
  make<B, E, R>((ctx) => {
    const result = fa.run(ctx)
    if (result._tag !== "Right") return result
    return makeInternalSuccess(f(result.value))
  }))

export const orElse = <E, B, E2, R2>(
  f: (e: E) => Nano<B, E2, R2>
) =>
<A, R>(fa: Nano<A, E, R>) =>
  make<A | B, E2, R | R2>((ctx) => {
    const result = fa.run(ctx)
    if (result._tag === "Left") return f(result.value).run(ctx)
    return result
  })

export const firstSuccessOf = <A extends Array<Nano<any, any, any>>>(
  arr: A
): Nano<A[number]["~nano.success"], A[number]["~nano.error"], A[number]["~nano.requirements"]> =>
  arr.slice(1).reduce((arr, fa) => orElse(() => fa)(arr), arr[0])

export const service = <I extends NanoTag<any>>(tag: I) =>
  make<I["~nano.requirements"], never, I["~nano.requirements"]>((ctx) =>
    (tag.key in ctx.value)
      ? (ctx.value[tag.key] as NanoExit<I["~nano.requirements"], never>)
      : makeInternalDefect(`Cannot find service ${tag.key}`)
  )

export const provideService = <I extends NanoTag<any>>(
  tag: I,
  value: I["~nano.requirements"]
) =>
<A, E, R>(fa: Nano<A, E, R>) =>
  make<A, E, Exclude<R, I["~nano.requirements"]>>((ctx) => {
    return fa.run({
      ...ctx,
      value: {
        ...ctx.value,
        [tag.key]: makeInternalSuccess(value)
      }
    })
  })

export const gen = <Eff extends Gen.YieldWrap<Nano<any, any, any>>, AEff>(
  ...args: [body: () => Generator<Eff, AEff, never>]
) =>
  make<
    AEff,
    [Eff] extends [never] ? never
      : [Eff] extends [Gen.YieldWrap<Nano<infer _A, infer E, infer _R>>] ? E
      : never,
    [Eff] extends [never] ? never
      : [Eff] extends [Gen.YieldWrap<Nano<infer _A, infer _E, infer R>>] ? R
      : never
  >((ctx) => {
    const iterator = args[0]()
    let state: IteratorResult<any> = iterator.next()
    while (!state.done) {
      const current = Gen.isGenKind(state.value)
        ? state.value.value
        : Gen.yieldWrapGet(state.value)
      const result: NanoExit<any, any> = current.run(ctx)
      if (result._tag !== "Right") {
        return result
      }
      state = iterator.next(result.value as never)
    }
    return makeInternalSuccess(state.value)
  })

export const fn = (_: string) =>
<Eff extends Gen.YieldWrap<Nano<any, any, any>>, AEff, Args extends Array<any>>(
  body: (...args: Args) => Generator<Eff, AEff, never>
) =>
(...args: Args) => (
  make<
    AEff,
    [Eff] extends [never] ? never
      : [Eff] extends [Gen.YieldWrap<Nano<infer _A, infer E, infer _R>>] ? E
      : never,
    [Eff] extends [never] ? never
      : [Eff] extends [Gen.YieldWrap<Nano<infer _A, infer _E, infer R>>] ? R
      : never
  >((ctx) => {
    const iterator = body(...args)
    let state: IteratorResult<any> = iterator.next()
    while (!state.done) {
      const current = Gen.isGenKind(state.value)
        ? state.value.value
        : Gen.yieldWrapGet(state.value)
      const result: NanoExit<any, any> = current.run(ctx)
      if (result._tag !== "Right") {
        return result
      }
      state = iterator.next(result.value as never)
    }
    return makeInternalSuccess(state.value)
  })
)

export const option = <A, E, R>(fa: Nano<A, E, R>) =>
  make<Option.Option<A>, never, R>((ctx) => {
    const result = fa.run(ctx)
    switch (result._tag) {
      case "Right":
        return makeInternalSuccess(Option.some(result.value))
      case "Left":
        return makeInternalSuccess(Option.none())
      case "Defect":
        return result
    }
  })

const successUndefined = makeInternalSuccess(undefined)

export const void_ = make<void, never, never>(() => successUndefined)

export const ignore = <A, E, R>(fa: Nano<A, E, R>) =>
  make<void, never, R>((ctx) => {
    fa.run(ctx)
    return successUndefined
  })

export const all = <A extends Array<Nano<any, any, any>>>(
  ...args: A
): Nano<
  Array<A[number]["~nano.success"]>,
  A[number]["~nano.error"],
  A[number]["~nano.requirements"]
> =>
  make<
    Array<A[number]["~nano.success"]>,
    A[number]["~nano.error"],
    A[number]["~nano.requirements"]
  >((ctx) => {
    const results: Array<A[number]["~nano.success"]> = []
    for (const arg of args) {
      const result = arg.run(ctx)
      if (result._tag !== "Right") return result
      results.push(result.value)
    }
    return makeInternalSuccess(results)
  })

const timings: Record<string, number> = {}
const timingsCount: Record<string, number> = {}
export const timed = (timingName: string) => <A, E, R>(fa: Nano<A, E, R>) =>
  make<A, E, R>((ctx) => {
    const start = performance.now()
    const result = fa.run(ctx)
    const end = performance.now()
    const duration = end - start
    timings[timingName] = (timings[timingName] || 0) + duration
    timingsCount[timingName] = (timingsCount[timingName] || 0) + 1
    return result
  })

export const getTimings = () => {
  const result: Array<[name: string, avg: number, hits: number, total: number]> = []
  for (const key in timings) {
    result.push([key, timings[key] / (timingsCount[key] || 1), timingsCount[key], timings[key]])
  }
  result.sort((a, b) => b[3] - a[3])
  const lines: Array<string> = []
  for (const [name, avg, hits, total] of result) {
    lines.push(
      `${name.padEnd(75)} tot ${total.toFixed(2).padStart(10)}ms avg ${avg.toFixed(2).padStart(10)}ms ${
        hits.toString().padStart(10)
      } hits`
    )
  }
  return lines
}

const internalNanoCache = Tag<Record<string, Map<any, NanoExit<any, any>>>>(
  "@effect/language-service/internalNanoCache"
)

export function cachedBy<P extends Array<any>, A, E, R>(
  fa: (...args: P) => Nano<A, E, R>,
  key: string,
  lookupKey: (...args: P) => any
) {
  return (...args: P) =>
    make<A, E, R>((ctx) => {
      const cacheObj = ctx.value[internalNanoCache.key] as Record<string, Map<any, NanoExit<any, any>>>
      const cache = cacheObj[key] || new Map<any, NanoExit<any, any>>()
      cacheObj[key] = cache
      const lookup = lookupKey(...args)
      if (cache.has(lookup)) {
        return cache.get(lookup)!
      }
      const result = fa(...args).run(ctx)
      cache.set(lookup, result)
      return result
    })
}
