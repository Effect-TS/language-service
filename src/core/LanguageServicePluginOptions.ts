import { isArray } from "effect/Array"
import * as Array from "effect/Array"
import { pipe } from "effect/Function"
import { hasProperty, isBoolean, isNumber, isObject, isRecord, isString } from "effect/Predicate"
import * as Record from "effect/Record"
import * as Nano from "./Nano"

export type DiagnosticSeverity = "error" | "warning" | "message" | "suggestion"

export interface LanguageServicePluginOptions {
  refactors: boolean
  diagnostics: boolean
  diagnosticSeverity: Record<string, DiagnosticSeverity | "off">
  quickinfoEffectParameters: "always" | "never" | "whentruncated"
  quickinfo: boolean
  quickinfoMaximumLength: number
  completions: boolean
  goto: boolean
  inlays: boolean
  allowedDuplicatedPackages: Array<string>
  namespaceImportPackages: Array<string>
  topLevelNamedReexports: "ignore" | "follow"
  barrelImportPackages: Array<string>
  importAliases: Record<string, string>
  renames: boolean
  noExternal: boolean
}

export const LanguageServicePluginOptions = Nano.Tag<LanguageServicePluginOptions>("PluginOptions")

function parseDiagnosticSeverity(config: Record<PropertyKey, unknown>): Record<string, DiagnosticSeverity | "off"> {
  if (!isRecord(config)) return {}
  return Object.fromEntries(
    pipe(
      Object.entries(config),
      Array.filter(([key, value]) => isString(key) && isString(value)),
      Array.map(([key, value]) => [String(key).toLowerCase(), String(value).toLowerCase()]),
      Array.filter(([_, value]) =>
        value === "off" || value === "error" || value === "warning" || value === "message" || value === "suggestion"
      )
    )
  )
}

export const defaults: LanguageServicePluginOptions = {
  refactors: true,
  diagnostics: true,
  diagnosticSeverity: {},
  quickinfo: true,
  quickinfoEffectParameters: "whentruncated",
  quickinfoMaximumLength: -1,
  completions: true,
  goto: true,
  inlays: true,
  allowedDuplicatedPackages: [],
  namespaceImportPackages: [],
  topLevelNamedReexports: "ignore",
  barrelImportPackages: [],
  importAliases: {},
  renames: true,
  noExternal: false
}

export function parse(config: any): LanguageServicePluginOptions {
  return {
    refactors: isObject(config) && hasProperty(config, "refactors") && isBoolean(config.refactors)
      ? config.refactors
      : defaults.refactors,
    diagnostics: isObject(config) && hasProperty(config, "diagnostics") && isBoolean(config.diagnostics)
      ? config.diagnostics
      : defaults.diagnostics,
    diagnosticSeverity:
      isObject(config) && hasProperty(config, "diagnosticSeverity") && isRecord(config.diagnosticSeverity)
        ? parseDiagnosticSeverity(config.diagnosticSeverity)
        : defaults.diagnosticSeverity,
    quickinfo: isObject(config) && hasProperty(config, "quickinfo") && isBoolean(config.quickinfo)
      ? config.quickinfo
      : defaults.quickinfo,
    quickinfoEffectParameters: isObject(config) && hasProperty(config, "quickinfoEffectParameters") &&
        isString(config.quickinfoEffectParameters) &&
        ["always", "never", "whentruncated"].includes(config.quickinfoEffectParameters.toLowerCase())
      ? config.quickinfoEffectParameters.toLowerCase() as "always" | "never" | "whentruncated"
      : defaults.quickinfoEffectParameters,
    quickinfoMaximumLength:
      isObject(config) && hasProperty(config, "quickinfoMaximumLength") && isNumber(config.quickinfoMaximumLength)
        ? config.quickinfoMaximumLength
        : defaults.quickinfoMaximumLength,
    completions: isObject(config) && hasProperty(config, "completions") && isBoolean(config.completions)
      ? config.completions
      : defaults.completions,
    goto: isObject(config) && hasProperty(config, "goto") && isBoolean(config.goto)
      ? config.goto
      : defaults.goto,
    inlays: isObject(config) && hasProperty(config, "inlays") && isBoolean(config.inlays)
      ? config.inlays
      : defaults.inlays,
    allowedDuplicatedPackages: isObject(config) && hasProperty(config, "allowedDuplicatedPackages") &&
        isArray(config.allowedDuplicatedPackages) && config.allowedDuplicatedPackages.every(isString)
      ? config.allowedDuplicatedPackages.map((_) => _.toLowerCase())
      : defaults.allowedDuplicatedPackages,
    namespaceImportPackages: isObject(config) && hasProperty(config, "namespaceImportPackages") &&
        isArray(config.namespaceImportPackages) && config.namespaceImportPackages.every(isString)
      ? config.namespaceImportPackages.map((_) => _.toLowerCase())
      : defaults.namespaceImportPackages,
    barrelImportPackages: isObject(config) && hasProperty(config, "barrelImportPackages") &&
        isArray(config.barrelImportPackages) && config.barrelImportPackages.every(isString)
      ? config.barrelImportPackages.map((_) => _.toLowerCase())
      : defaults.barrelImportPackages,
    importAliases: isObject(config) && hasProperty(config, "importAliases") && isRecord(config.importAliases)
      ? Record.map(config.importAliases, (value) => String(value))
      : defaults.importAliases,
    topLevelNamedReexports: isObject(config) && hasProperty(config, "topLevelNamedReexports") &&
        isString(config.topLevelNamedReexports) &&
        ["ignore", "follow"].includes(config.topLevelNamedReexports.toLowerCase())
      ? config.topLevelNamedReexports.toLowerCase() as "ignore" | "follow"
      : defaults.topLevelNamedReexports,
    renames: isObject(config) && hasProperty(config, "renames") && isBoolean(config.renames)
      ? config.renames
      : defaults.renames,
    noExternal: isObject(config) && hasProperty(config, "noExternal") && isBoolean(config.noExternal)
      ? config.noExternal
      : defaults.noExternal
  }
}
