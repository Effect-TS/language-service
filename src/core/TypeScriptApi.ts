import { hasProperty, isObject, isString } from "effect/Predicate"
import type ts from "typescript"
import * as Nano from "../core/Nano.js"

declare module "typescript" {
  const nullTransformationContext: ts.TransformationContext

  export namespace formatting {
    interface FormattingHost {
      getNewLine?(): string
    }

    export interface FormatContext {
      readonly options: ts.FormatCodeSettings
      readonly getRules: unknown
    }

    function getFormatContext(options: ts.FormatCodeSettings, host: FormattingHost): FormatContext
  }

  export type TextChangesContext = any

  export namespace textChanges {
    export interface ChangeNodeOptions extends ConfigurableStartEnd, InsertNodeOptions {}
    export enum LeadingTriviaOption {
      /** Exclude all leading trivia (use getStart()) */
      Exclude = 0,
      /** Include leading trivia and,
       * if there are no line breaks between the node and the previous token,
       * include all trivia between the node and the previous token
       */
      IncludeAll = 1,
      /**
       * Include attached JSDoc comments
       */
      JSDoc = 2,
      /**
       * Only delete trivia on the same line as getStart().
       * Used to avoid deleting leading comments
       */
      StartLine = 3
    }
    export enum TrailingTriviaOption {
      /** Exclude all trailing trivia (use getEnd()) */
      Exclude = 0,
      /** Doesn't include whitespace, but does strip comments */
      ExcludeWhitespace = 1,
      /** Include trailing trivia */
      Include = 2
    }
    export interface ConfigurableStart {
      leadingTriviaOption?: LeadingTriviaOption
    }
    export interface ConfigurableEnd {
      trailingTriviaOption?: TrailingTriviaOption
    }
    export interface InsertNodeOptions {
      /**
       * Text to be inserted before the new node
       */
      prefix?: string
      /**
       * Text to be inserted after the new node
       */
      suffix?: string
      /**
       * Text of inserted node will be formatted with this indentation, otherwise indentation will be inferred from the old node
       */
      indentation?: number
      /**
       * Text of inserted node will be formatted with this delta, otherwise delta will be inferred from the new node kind
       */
      delta?: number
    }
    export interface ConfigurableStartEnd extends ConfigurableStart, ConfigurableEnd {
    }
    export class ChangeTracker {
      static with(
        context: ts.TextChangesContext,
        cb: (tracker: ChangeTracker) => void
      ): Array<ts.FileTextChanges>
      delete(
        sourceFile: ts.SourceFile,
        node: ts.Node | ts.NodeArray<ts.TypeParameterDeclaration>
      ): void
      deleteRange(sourceFile: ts.SourceFile, range: ts.TextRange): void
      replaceNode(
        sourceFile: ts.SourceFile,
        oldNode: ts.Node,
        newNode: ts.Node,
        options?: ts.textChanges.ChangeNodeOptions
      ): void
      insertNodeAt(
        sourceFile: ts.SourceFile,
        pos: number,
        newNode: ts.Node,
        options?: ts.textChanges.InsertNodeOptions
      ): void
      insertNodeBefore(
        sourceFile: ts.SourceFile,
        before: ts.Node,
        newNode: ts.Node,
        blankLineBetween?: boolean,
        options?: ConfigurableStartEnd
      ): void
      insertNodeAfter(sourceFile: ts.SourceFile, after: ts.Node, newNode: ts.Node): void
      insertText(sourceFile: ts.SourceFile, pos: number, text: string): void
      insertCommentBeforeLine(
        sourceFile: ts.SourceFile,
        lineNumber: number,
        position: number,
        commentText: string
      ): void
    }
    export function applyChanges(text: string, changes: ReadonlyArray<ts.TextChange>): string
  }

  export function findPrecedingToken(
    position: number,
    sourceFile: ts.SourceFileLike,
    startNode: ts.Node,
    excludeJsdoc?: boolean
  ): ts.Node | undefined
  export function findPrecedingToken(
    position: number,
    sourceFile: ts.SourceFile,
    startNode?: ts.Node,
    excludeJsdoc?: boolean
  ): ts.Node | undefined
  function findChildOfKind<T extends ts.Node>(
    n: ts.Node,
    kind: T["kind"],
    sourceFile: ts.SourceFileLike
  ): T | undefined

  export function isMemberName(node: ts.Node): node is ts.MemberName
  export function isKeyword(token: ts.SyntaxKind): token is ts.KeywordSyntaxKind

  export interface TypeChecker {
    isTypeAssignableTo(source: ts.Type, target: ts.Type): boolean
    getUnionType(types: ReadonlyArray<ts.Type>): ts.Type
  }

  export function typeToDisplayParts(
    typechecker: ts.TypeChecker,
    type: ts.Type,
    enclosingDeclaration?: ts.Node | undefined,
    flags?: ts.TypeFormatFlags
  ): Array<ts.SymbolDisplayPart>
}

type _TypeScriptApi = typeof ts
export interface TypeScriptApi extends _TypeScriptApi {}
export const TypeScriptApi = Nano.Tag<TypeScriptApi>("TypeScriptApi")

type _TypeScriptProgram = ts.Program
export interface TypeScriptProgram extends _TypeScriptProgram {}
export const TypeScriptProgram = Nano.Tag<TypeScriptProgram>("TypeScriptProgram")

export const ChangeTracker = Nano.Tag<ts.textChanges.ChangeTracker>("ChangeTracker")

interface ModuleWithPackageInfo {
  name: string
  version: string
  hasEffectInPeerDependencies: boolean
  contents: any
  packageDirectory: string
}

export function parsePackageContentNameAndVersionFromScope(v: unknown): ModuleWithPackageInfo | undefined {
  if (!isObject(v)) return
  if (!hasProperty(v, "packageJsonScope")) return
  if (!v.packageJsonScope) return
  const packageJsonScope = v.packageJsonScope
  if (!hasProperty(packageJsonScope, "contents")) return
  if (!hasProperty(packageJsonScope.contents, "packageJsonContent")) return
  const packageJsonContent = packageJsonScope.contents.packageJsonContent
  if (!hasProperty(packageJsonContent, "name")) return
  if (!hasProperty(packageJsonContent, "version")) return
  if (!hasProperty(packageJsonScope, "packageDirectory")) return
  if (!isString(packageJsonScope.packageDirectory)) return
  const { name, version } = packageJsonContent
  if (!isString(name)) return
  if (!isString(version)) return
  const hasEffectInPeerDependencies = hasProperty(packageJsonContent, "peerDependencies") &&
    isObject(packageJsonContent.peerDependencies) &&
    hasProperty(packageJsonContent.peerDependencies, "effect")
  return {
    name: name.toLowerCase(),
    version: version.toLowerCase(),
    hasEffectInPeerDependencies,
    contents: packageJsonContent,
    packageDirectory: packageJsonScope.packageDirectory
  }
}
