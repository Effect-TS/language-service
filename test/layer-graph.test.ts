import * as LanguageServicePluginOptions from "@effect/language-service/core/LanguageServicePluginOptions"
import * as LayerGraph from "@effect/language-service/core/LayerGraph"
import * as Nano from "@effect/language-service/core/Nano"
import * as TypeCheckerApi from "@effect/language-service/core/TypeCheckerApi"
import * as TypeCheckerUtils from "@effect/language-service/core/TypeCheckerUtils"
import * as TypeParser from "@effect/language-service/core/TypeParser"
import * as TypeScriptApi from "@effect/language-service/core/TypeScriptApi"
import * as TypeScriptUtils from "@effect/language-service/core/TypeScriptUtils"
import * as Either from "effect/Either"
import { pipe } from "effect/Function"
import * as fs from "fs"
import * as path from "path"
import * as ts from "typescript"
import { describe, expect, it } from "vitest"
import { configFromSourceComment, createServicesWithMockedVFS } from "./utils/mocks.js"

const getExamplesLayerGraph = () => path.join(__dirname, "..", "examples", "layer-graph")

async function testLayerGraphOnExample(fileName: string, sourceText: string) {
  const { program, sourceFile } = createServicesWithMockedVFS(
    fileName,
    sourceText
  )

  // get the node to test
  const nodes: Array<[name: string, node: ts.Node]> = []
  for (const statement of sourceFile.statements) {
    if (
      ts.isVariableStatement(statement) && statement.modifiers &&
      statement.modifiers.some((_) => _.kind === ts.SyntaxKind.ExportKeyword)
    ) {
      for (const declaration of statement.declarationList.declarations) {
        if (ts.isVariableDeclaration(declaration) && ts.isIdentifier(declaration.name) && declaration.initializer) {
          nodes.push([ts.idText(declaration.name), declaration.initializer])
        }
      }
    }
  }

  // at least one test
  expect(nodes.length).toBeGreaterThan(0)
  expect(nodes.map(([name]) => name)).toMatchSnapshot()

  // loop and test the nodes
  for (const [name, node] of nodes) {
    // create snapshot path
    const baseSnapshotFilePath = path.join(
      __dirname,
      "__snapshots__",
      "layer-graph",
      fileName + "_" + name
    )

    const options = LanguageServicePluginOptions.parse({
      ...LanguageServicePluginOptions.defaults,
      completions: true,
      refactors: false,
      diagnostics: false,
      quickinfo: false,
      goto: false,
      ...configFromSourceComment(sourceText)
    })

    // check and assert the completions is executable
    const maybeGraph = pipe(
      Nano.gen(function*() {
        const layerGraph = yield* LayerGraph.extractLayerGraph(node, {
          arrayLiteralAsMerge: false,
          explodeOnlyLayerCalls: false,
          followSymbolsDepth: options.layerGraphFollowDepth
        })
        const outlineGraph = yield* LayerGraph.extractOutlineGraph(layerGraph)
        const providersAndRequirers = yield* LayerGraph.extractProvidersAndRequirers(layerGraph)
        return {
          layerGraph: yield* LayerGraph.formatLayerGraph(layerGraph, sourceFile),
          layerNestedGraph: yield* LayerGraph.formatNestedLayerGraph(layerGraph, sourceFile),
          outlineGraph: yield* LayerGraph.formatLayerOutlineGraph(outlineGraph, sourceFile),
          providersAndRequirers: yield* LayerGraph.formatLayerProvidersAndRequirersInfo(
            providersAndRequirers,
            sourceFile
          )
        }
      }),
      TypeParser.nanoLayer,
      TypeCheckerUtils.nanoLayer,
      TypeScriptUtils.nanoLayer,
      Nano.provideService(TypeCheckerApi.TypeCheckerApi, program.getTypeChecker()),
      Nano.provideService(TypeScriptApi.TypeScriptProgram, program),
      Nano.provideService(TypeScriptApi.TypeScriptApi, ts),
      Nano.provideService(
        LanguageServicePluginOptions.LanguageServicePluginOptions,
        options
      ),
      Nano.unsafeRun
    )

    if (Either.isLeft(maybeGraph)) {
      await expect("no graph").toMatchFileSnapshot(baseSnapshotFilePath + ".output")
      return
    }

    await expect(maybeGraph.right.layerGraph, "layerGraph").toMatchFileSnapshot(baseSnapshotFilePath + ".output")
    await expect(maybeGraph.right.layerNestedGraph, "layerNestedGraph").toMatchFileSnapshot(
      baseSnapshotFilePath + ".nested"
    )
    await expect(maybeGraph.right.outlineGraph, "outlineGraph").toMatchFileSnapshot(baseSnapshotFilePath + ".outline")
    await expect(maybeGraph.right.providersAndRequirers, "providersAndRequirers").toMatchFileSnapshot(
      baseSnapshotFilePath + ".quickinfo"
    )
  }
}

function testAllCompletions() {
  // read all filenames
  const allExampleFiles = fs.readdirSync(getExamplesLayerGraph())

  describe("Layer Graph ", () => {
    // for each example file
    for (const fileName of allExampleFiles) {
      it(fileName, async () => {
        const sourceText = fs.readFileSync(path.join(getExamplesLayerGraph(), fileName))
          .toString("utf8")
        await testLayerGraphOnExample(fileName, sourceText)
      })
    }
  })
}

testAllCompletions()
