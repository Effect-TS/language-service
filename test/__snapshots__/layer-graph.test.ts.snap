// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Layer Graph  > duplicated.ts 1`] = `
[
  "AppLive",
]
`;

exports[`Layer Graph  > generics.ts 1`] = `
[
  "NoComment",
]
`;

exports[`Layer Graph  > multiple.ts 1`] = `
[
  "AppLive",
]
`;

exports[`Layer Graph  > simple.ts 1`] = `
[
  "expect",
  "simplePipeIn",
  "liveWithPipeable",
]
`;

exports[`Layer Graph  > specialChars.ts 1`] = `
[
  "NoComment",
]
`;
